--- FunsExternal_Template.lean
+++ FunsExternal.lean
@@ -74,10 +74,14 @@
     Name pattern: [core::result::{core::result::Result<@T, @E>}::map]
     Visibility: public -/
 @[rust_fun "core::result::{core::result::Result<@T, @E>}::map"]
-axiom core.result.Result.map
-  {T : Type} {E : Type} {U : Type} {F : Type} (opsfunctionFnOnceFTupleTUInst :
+def core.result.Result.map
+  {T : Type} {E : Type} {U : Type} {F : Type} (inst :
   core.ops.function.FnOnce F T U) :
   core.result.Result T E → F → Result (core.result.Result U E)
+  | core.result.Result.Ok v, x => do
+      let u ← inst.call_once x v
+      ok (core.result.Result.Ok u)
+  | core.result.Result.Err e, _ => ok (core.result.Result.Err e)
 
 /-- [core::result::{core::ops::try_trait::Try<T, core::result::Result<core::convert::Infallible, E>> for core::result::Result<T, E>}::branch]:
     Source: '/rustc/library/core/src/result.rs', lines 2172:4-2172:64
@@ -85,10 +89,14 @@
     Visibility: public -/
 @[rust_fun
   "core::result::{core::ops::try_trait::Try<core::result::Result<@T, @E>, @T, core::result::Result<core::convert::Infallible, @E>>}::branch"]
-axiom core.result.Result.Insts.CoreOpsTry_traitTryTResultInfallibleE.branch
+def core.result.Result.Insts.CoreOpsTry_traitTryTResultInfallibleE.branch
   {T : Type} {E : Type} :
   core.result.Result T E → Result (core.ops.control_flow.ControlFlow
     (core.result.Result core.convert.Infallible E) T)
+  | core.result.Result.Ok x =>
+      ok (core.ops.control_flow.ControlFlow.Continue x)
+  | core.result.Result.Err e =>
+      ok (core.ops.control_flow.ControlFlow.Break (core.result.Result.Err e))
 
 /-- [core::result::{core::ops::try_trait::FromResidual<core::result::Result<core::convert::Infallible, E>> for core::result::Result<T, F>}::from_residual]:
     Source: '/rustc/library/core/src/result.rs', lines 2187:4-2187:70
@@ -96,11 +104,15 @@
     Visibility: public -/
 @[rust_fun
   "core::result::{core::ops::try_trait::FromResidual<core::result::Result<@T, @F>, core::result::Result<core::convert::Infallible, @E>>}::from_residual"]
-axiom
+def
   core.result.Result.Insts.CoreOpsTry_traitFromResidualResultInfallibleE.from_residual
   (T : Type) {E : Type} {F : Type} (convertFromInst : core.convert.From F E) :
   core.result.Result core.convert.Infallible E → Result (core.result.Result T
     F)
+  | core.result.Result.Err e => do
+      let e' ← convertFromInst.from_ e
+      ok (core.result.Result.Err e')
+  | core.result.Result.Ok h => nomatch h
 
 /-- [std::backtrace::{std::backtrace::Backtrace}::capture]:
     Source: '/rustc/library/std/src/backtrace.rs', lines 292:4-292:33
